### ISSUE TITLE
poudriere-bulk -S flag confusing behavior

### ISSUE THREAD

**Issue by valpackett** [freebsd/poudriere#806]
Looking at this:

```
     -S           Do not recursively rebuild packages affected by other packages requiring
                  incremental rebuild.  This may result in broken packages if the ones they depend
                  on are updated, are not ABI-compatible, and were not properly PORTREVISION bumped.
```

what I expect: when some commonly used dependency is updated, **it** gets rebuilt (*including as a dependency*), but it doesn't cause all consumers to be automatically rebuilt.

what happens: it doesn't get rebuilt as a dependency at all (when building something new (as in not-built-yet) that depends on it) and the build just proceeds without *any* version of that dependency, which is silly:

```
[00:01:04] Checking packages for incremental rebuild needs
[00:01:06] Deleting librsvg2-rust-2.50.0_2.tzst: new version: 2.50.2
[00:01:07] (-S) Skipping recursive rebuild
[…]
[00:04:34] [01] [00:03:22] Finished editors/libreoffice | libreoffice-7.0.3.1_2: Failed: lib-depends
```

log:

```
pkg-static: Missing dependency 'librsvg2-rust'

Failed to install the following 1 package(s): /packages/All/gtk3-3.24.23_2.tzst
```

**Comment by myfreeweb** [comment]
But then it's even more broken: I rebuild `librsvg2-rust`, it's there

```
% la /usr/local/poudriere/data/packages/dank-default/All/librsvg2-rust-2.50.2.tzst
-rw-r--r--  1 root  greg   3.4M Dec  5 23:26 /usr/local/poudriere/data/packages/dank-default/All/librsvg2-rust-2.50.2.tzst
```

I can upgrade it on the host system from the poudriere repo. But the `libreoffice` build fails still. Because the script that does the `===>   Installing existing package` thing when building *does not* install `librsvg2-rust` anymore!! >_<

**Comment by myfreeweb** [comment]
Aha!! Found the code responsible for excluding these packages, it is the `TRIM_ORPHANED_BUILD_DEPS` thing in `clean_build_queue()`.

here's the outputs of `${tmp}`, `${tmp}.actual` and `comm -13 ${tmp} ${tmp}.actual`, the latter going into `pkgqueue_remove_many_pipe`: 

```
[00:00:52] TMP
pkg-1.15.10_1
py39-scikit-image-0.17.2
[00:00:52] /TMP
[00:00:52] ACT
py39-scikit-image-0.17.2
py39-sphinx-3.3.1,1
[00:00:52] /ACT
[00:00:52] COMM
py39-sphinx-3.3.1,1
[00:00:52] /COMM
```

where `py39-sphinx` is indeed the transitive dependency that must be built.

I guess the documentation for the `-S` flag should just mention that the `TRIM_ORPHANED_BUILD_DEPS` option will lead to missing packages with it.

**Comment by myfreeweb** [comment]
Welp, unfortunately this still has a problem — it tries to build these packages in parallel.

```
[00:03:21] [01] [00:00:00] Building textproc/py-sphinx@py39 | py39-sphinx-3.3.1,1
[00:03:21] [02] [00:00:00] Building graphics/py-scikit-image@py39 | py39-scikit-image-0.17.2
[00:03:37] [01] [00:00:16] Finished textproc/py-sphinx@py39 | py39-sphinx-3.3.1,1: Success
[00:03:38] [02] [00:00:17] Finished graphics/py-scikit-image@py39 | py39-scikit-image-0.17.2: Failed: build-depends
```

So these deps really shouldn't be orphaned, they should ideally be like, **reparented**.

I do have a lazy workaround though: unsurprisingly, moving the pkg-already-exists check deeper into the build helps :)

```diff
--- common.sh.orig	2020-04-10 18:42:02.935984000 +0300
+++ common.sh	2020-12-06 17:31:07.843386000 +0300
@@ -4110,6 +4117,14 @@
 		COLOR_ARROW="${COLOR_IGNORE}" job_msg "${COLOR_IGNORE}Finished ${COLOR_PORT}${port}${FLAVOR:+@${FLAVOR}} | ${PKGNAME}${COLOR_IGNORE}: Ignored: ${ignore}"
 		clean_rdepends="ignored"
 		run_hook pkgbuild ignored "${port}" "${PKGNAME}" "${ignore}" >&3
+	elif [ -f "${PACKAGES}/All/${PKGNAME}.${PKG_EXT}" ]; then
+		msg "Already built ${port}"
+		badd ports.built "${ORIGINSPEC} ${PKGNAME} ${elapsed}"
+		COLOR_ARROW="${COLOR_SUCCESS}" job_msg "${COLOR_SUCCESS}Finished ${COLOR_PORT}${port}${FLAVOR:+@${FLAVOR}} | ${PKGNAME}${COLOR_SUCCESS}: Already built"
+		run_hook pkgbuild success "${port}" "${PKGNAME}" >&3
+		# Cache information for next run
+		pkg_cacher_queue "${port}" "${pkgname}" \
+		    "${DEPENDS_ARGS}" "${FLAVOR}" || :
 	else
 		build_port "${ORIGINSPEC}" || ret=$?
 		if [ ${ret} -ne 0 ]; then
@@ -6922,9 +6937,9 @@
 	msg "Cleaning the build queue"
 
 	# Delete from the queue all that already have a current package.
-	pkgqueue_list | while mapfile_read_loop_redir pn; do
-		[ -f "../packages/All/${pn}.${PKG_EXT}" ] && echo "${pn}"
-	done | pkgqueue_remove_many_pipe
+	# pkgqueue_list | while mapfile_read_loop_redir pn; do
+	# 	[ -f "../packages/All/${pn}.${PKG_EXT}" ] && echo "${pn}"
+	# done | pkgqueue_remove_many_pipe
 
 	# Delete from the queue orphaned build deps. This can happen if
 	# the specified-to-build ports have all their deps satisifed
```

at the cost of more annoying log output and the overhead of starting-stopping builders

**Comment by bdrewery** [comment]
I too just discovered that `-S` does not play well with `TRIM_ORPHANED_BUILD_DEPS`.

1. `delete_old_pkgs`: If a package gets a new version (or other legit rebuild needs) it is deleted but kept in the queue.
2. `sanity_check_pkgs` (without `-S`): packages which have missing dependencies are deleted but kept in the queue.
3. `clean_queue`: If a package exists after the recursive checks it is unqueued.
4. `clean_queue` second half w/`TRIM_ORPHANED_BUILD_DEPS`: If a queued build is not a "listed" port (in `bulk -f`) and nothing depends on it then it is unqueued. This incorrectly deletes runtime dependencies. I don't think there would be a problem if it only deleted build dependencies.

I will have a prototype in the next poudriere-devel that supports this without using `-S` (at least in its current form) (#822)
What we need is for `pkg` to not write the versions into the dependencies (https://github.com/freebsd/pkg/issues/1831 -> `env PKG_NO_VERSION_FOR_DEPS=1 make package`). Then over time, **without `-S`** the old packages with versions in them will be deleted via `sanity_check_pkgs` and rebuilt with the new dependencies not containing versions. Then those specific packages won't get deleted by `sanity_check_pkgs` again in the future as (in #822) it automatically excludes these new packages from being recursively deleted. Simply using `-S` would require a full 1 time rebuild of the repository.I have not yet decided how it will be enabled but it might just be via `-S` initially and later enabled by default.

Closing for #822.

**Comment by bdrewery** [comment]
Reopening as this is a major change/fix in relation to #822 that I want to track separately. The solution I came up with is effectively the same as https://github.com/freebsd/poudriere/issues/806#issuecomment-739513867 and removing `TRIM_ORPHANED_BUILD_DEPS`.
